### PR TITLE
Caching - Full refactoring.

### PR DESCRIPTION
**Pull request by leogrim**

### CODE REVIEW

File: shoebox/modules/common/app/com/keepit/model/BrowsingHistory.scala [hunk 1]
```diff
@@ -49,25 +49,8 @@ case class BrowsingHistoryUserIdKey(userId: Id[User]) extends Key[BrowsingHistor
...
-@Singleton
-class BrowsingHistoryUserIdCacheImpl @Inject() (
-    val repo: InMemoryCachePlugin,
-    outerCacheImpl: OuterBrowsingHistoryUserIdCacheImpl)
-  extends {
-    val ttl = 10 seconds
-    override val outerCache = Some(outerCacheImpl)
-  } with BrowsingHistoryUserIdCache
-
+class BrowsingHistoryUserIdCache @Inject() (innerRepo: InMemoryCachePlugin, outerRepo: FortyTwoCachePlugin)
```
Comment (eishay) [inline shoebox/modules/common/app/com/keepit/model/BrowsingHistory.scala:53]: whoo, where did all this boilerplate code went? :-)

File: shoebox/modules/common/app/com/keepit/common/cache/FortyTwoCache.scala [hunk 2]
```diff
@@ -299,7 +297,48 @@ trait FortyTwoCache[K <: Key[T], T] extends ObjectCache[K, T] {
...
+
+  // Constructor using the same serializer for each cache plugin
+  def this(innermostPlugin: (FortyTwoCachePlugin, Duration), innerToOuterPlugins: (FortyTwoCachePlugin, Duration)*)(serializer: Serializer[T]) =
+    this(innermostPlugin._1, innermostPlugin._2, serializer, FortyTwoCacheFactory[K, T](innerToOuterPlugins)(serializer))
 
-  def parseJson(obj: Any)(implicit formatter: Format[T]): T =
-    Json.fromJson[T](Json.parse(obj.asInstanceOf[String]).asInstanceOf[JsValue]).get
 }
+
+class JsonCacheImpl[K <: Key[T], T](innermostPlugin: (FortyTwoCachePlugin, Duration), innerToOuterPlugins: (FortyTwoCachePlugin, Duration)*)(implicit formatter: Format[T])
```
Comment (eishay) [inline shoebox/modules/common/app/com/keepit/common/cache/FortyTwoCache.scala:333]: nice!

**Comment by andrewconner** [comment]
Going through this now. Looks really great. The usage is much cleaner. :)

**Comment by andrewconner** [comment]
I think this is great. Do we need to flush the current cache, or should everything work the same?

File: shoebox/modules/common/app/com/keepit/serializer/Serializer.scala [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+object BinaryFormat {
+  implicit val clickHistoryBinaryFormat = ClickHistoryBinarySerializer.clickHistoryBinarySerializer
+  implicit val sliderHistoryBinaryFormat = SliderHistoryBinarySerializer.sliderHistoryBinarySerializer
+  implicit val browsingHistoryBinaryFormat = BrowsingHistoryBinarySerializer.browsingHistoryBinarySerializer
+}
+
+object TraversableFormat {
+  private def materialize[U, T <: Traversable[U]](implicit traversableFormatter: Format[T]) = traversableFormatter
+  def seq[U](implicit formatter: Format[U]) = materialize[U, Seq[U]]
```
Comment (2is10) [inline shoebox/modules/common/app/com/keepit/serializer/Serializer.scala:18]: poof!  right out of thin air.  and for my next trick...

**Comment by 2is10** [comment]
impressive

**Comment by leogrim** [comment]
Most of the caches should work the same, except maybe for the ones that
store IDs. Custom long serializers were used, I switched to the (default)
JSON formatter that was defined in the ID class. It outputs a jsNumber but
still reads a long so it may still work with the current content of the
caches.

But I guess it's just safer to flush anyway?

2013/6/10 Andrew Conner notifications@github.com

File: shoebox/modules/common/app/com/keepit/serializer/Serializer.scala (hunk 3 above)
Comment (leogrim) [inline shoebox/modules/common/app/com/keepit/serializer/Serializer.scala:18]: Hopefully, they'll have implicit bunnies in Scala 2.11.

2013/6/10 Jared Jacobs notifications@github.com

**Pull request merged**